cpu/stm32/periph/spi_u5.c

```diff
@@ -0,0 +1,386 @@
+  /* When SPE = 1, SPI data transfer is enabled, SPI_CFG1 and SPI_CFG2
+     configuration registers, CRCPOLY, UDRDR, part of SPI_AUTOCR register and
+     IOLOCK bit in the
+     SPI_CR1 register are write protected */
+}
+
+void spi_release(spi_t bus) {
+  /* disable device and release lock */
+  disable_spi(bus);
+  dev(bus)->CR2 = SPI_CR2_SETTINGS; /* Clear the DMA and SSOE flags */
+  dev(bus)->CFG2 = 0;
+  periph_clk_dis(spi_config[bus].apbbus, spi_config[bus].rccmask);
+#ifdef STM32_PM_STOP
+  /* unblock STOP mode */
+  pm_unblock(STM32_PM_STOP);
+#endif
+  mutex_unlock(&locks[bus]);
+}
+
+static inline void _wait_for_end(spi_t bus) {
+  /* make sure the transfer is completed before continuing, see reference
+   * manual(s) -> section 'Disabling the SPI' */
+  /* When TSIZE>0, EOT and TXC signals are equal so polling of EOT is reliable
+  at whatever SPI communication mode to check end of the bus activity */
+  while (!(dev(bus)->SR & SPI_SR_EOT)) {
+  }
+  while (dev(bus)->SR & SPI_SR_EOT) {
+    dev(bus)->IFCR = SPI_IFCR_EOTC;
+  }
+  dev(bus)->IFCR = SPI_IFCR_TXTFC;
+}
+
+static void _transfer_no_dma(spi_t bus, const void *out, void *in, size_t len) {
+  // data to be sent tx
+  const uint8_t *outbuf = out;
+  // dest for data to receive rx
+  uint8_t *inbuf = in;
+
+  // lower 16 bits = TSIZE
+  // When these bits are changed by software, the SPI must be disabled. SPE = 0
+  dev(bus)->CR2 = (int16_t)len;
+
+  /* we need to recast the data register to uint_8 to force 8-bit access */
+  volatile uint8_t *TXDR = (volatile uint8_t *)&(dev(bus)->TXDR);
+  volatile uint8_t *RXDR = (volatile uint8_t *)&(dev(bus)->RXDR);
+
+  /*To restart the internal
+    state machine properly, SPI is strongly suggested to be disabled and
+    re-enabled before next transaction starts despite its setting is not
+    changed.*/
+  disable_spi(bus);
+  enable_spi(bus);
+
+  if (dev(bus)->CFG2 & SPI_CFG2_MASTER) {
+    assume(dev(bus)->CR1 & SPI_CR1_SPE);
+    // CSTART: master transfer start This bit can be set by software if SPI is
+    // *enabled* only
+    dev(bus)->CR1 |= SPI_CR1_CSTART;
+    assume(dev(bus)->CR1 & SPI_CR1_CSTART);
+  }
+
+  /* transfer data, use shortpath if only sending data */
+  if (!inbuf) {
+    for (size_t i = 0; i < len; i++) {
+      while (!(dev(bus)->SR & SPI_SR_TXP)) {
+      }
+      *TXDR = outbuf[i];
+      while ((dev(bus)->SR & SPI_SR_RXP)) { /* busy wait */
+                                            // discard incoming bytes
+        (void)*RXDR;
+      }
+    }
+  } else if (!outbuf) {
+    for (size_t i = 0; i < len; i++) {
+      while (!(dev(bus)->SR & SPI_SR_TXP)) { /* busy wait */
+      }
+      *TXDR = 0;
+      // The RXP is cleared as soon as less data than complete single packet is
+      // available in the RxFIFO
+      while (!(dev(bus)->SR & SPI_SR_RXP)) { /* busy wait */
+      }
+      inbuf[i] = *RXDR;
+    }
+  } else {
+    volatile uint16_t tx_remainder = len, rx_remainder = len;
```

Comment:
```suggestion
    uint16_t tx_remainder = len, rx_remainder = len;
```

Also these should be on separate lines, I believe.